Integrations/integration-MCAS.yml

```diff
@@ -0,0 +1,265 @@
+  version: -1
+name: Microsoft Cloud App Security
+display: Microsoft Cloud App Security
+category: Network Security
+image: data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAbcAAAG3BAMAAADFuQbYAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAAElBMVEXz8/PzUyWBvAYFpvD/ugj///98Xc9SAAAAAWJLR0QF+G/pxwAAAAd0SU1FB+IECxAWCcy7n+sAAAGSSURBVHja7c9BCQAgEABBxQJawQSCFeyfyZcRTu4xW2CZUiRJkiRJylodwfV3mtHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcHBwcF9wUmSJEmSlK62g1vvdKKDg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4P7gpMkSZIkKUsX5n/TJVBjPXQAAAAldEVYdGRhdGU6Y3JlYXRlADIwMTgtMDQtMTFUMTY6MjI6MDkrMDA6MDD3dzx7AAAAJXRFWHRkYXRlOm1vZGlmeQAyMDE4LTA0LTExVDE2OjIyOjA5KzAwOjAwhiqExwAAAABJRU5ErkJggg==
+description: Microsoft Cloud App Security (MCAS)
+configuration:
+- display: Server URL
+  name: url
+  defaultvalue: ""
+  type: 0
+  required: true
+- display: API Token
+  name: token
+  defaultvalue: ""
+  type: 4
+  required: true
+- display: Ignore server certificate
+  name: insecure
+  defaultvalue: ""
+  type: 8
+  required: false
+- display: Use System Proxy
+  name: proxy
+  defaultvalue: ""
+  type: 8
+  required: false
+script:
+  script: |-
+    import requests
+    import json
+    import yaml
+
+    SERVER = demisto.params()['url'][:-1] if demisto.params()['url'].endswith('/') else demisto.params()['url']
+    TOKEN = demisto.params().get('token')
+    USE_SSL = not demisto.params().get('insecure', False)
+
+    if not demisto.params()['proxy']:
+        del os.environ['HTTP_PROXY']
+        del os.environ['HTTPS_PROXY']
+        del os.environ['http_proxy']
+        del os.environ['https_proxy']
+
+    def mcas_list_alerts():
+        limit = demisto.args()['limit']
+        severity_raw = demisto.args()['severity']
+        if severity_raw == "Low":
+            severity = "0"
+        elif severity_raw == "Medium":
+            severity = "1"
+        elif severity_raw == "High":
+            severity = "2"
+        else:
+            severity = "2,1,0"
+        url = SERVER + '/cas/api/v1/alerts/'
+        data = '{"skip":0,"limit":' + limit + ',"filters":{"resolutionStatus":{"eq":[0]},"severity":{"eq":[' + severity + ']}},"sortField":"date","sortDirection":"desc","performAsyncTotal":true}'
+        headers = {'Authorization': 'Token ' + TOKEN, 'Content-type': 'application/json'}
+        r = requests.post(url=url, headers=headers, data=data, verify=USE_SSL)
+        j = yaml.load(json.dumps(r.json()))
+        x = (len(j['data']))
+        results = []
+        demisto.results(j)
+        print('Number of fields: ' + str(x))
+        for i in range(0, x):
+            ip = "unknown"
```

Comment:
Not sure if we need these `unknown` defaults, why not get the value from the API and not do anything with it if it's not there?

Reply:
Each query results may or may not contain the value IP. Since each record in the response is being added to context, if I were to ignore fields that were not present, there would need to be multiple versions of the 'results' array, or "null" values in the context. Setting defaults to 'unknown' ended up resulting in much less code and makes playbook construction easier since there is always a string present in every context label. Should I still restructure?

Reply:
Is `unknown` also a value returned by the API, or a custom default value you set?

Reply:
It is a custom value. If I don't set the default as a string then the context will contain a `null` value for some records.

Reply:
So it is better not to return it and make sure to remove the `null`
We have builtin functions for that:
For human readable, you can pass `removeNull=True` to `tableToMarkdown`.
Same argument exist for `createContext`

Reply:
Thanks @Itay4. `removeNull=True` is a great mention -- I didn't realize that was an option. However, I still need to assign the default string value as the json fields vary based on the type of alert returned. For instance, 5 different alert types may have IP, description, username, and time, but 3 other ones may have username, time, location, and reputation. If I only assign the set the variable when the field is present, the next iteration of the loop will have the previous value for IP instead of being null. Now, I could fully mitigate this will different case statements for each alert type, but since there are so many alert types this is much more practical. What are your thoughts?

Reply:
Let's schedule an execution review and go over this together and see what would be the right thing to do here.
Will reach out to you in DFIR.